src/Processors/Transforms/JoiningTransform.cpp

```diff
@@ -327,12 +327,48 @@ void FillingRightJoinSideTransform::work()
     else
         stop_reading = !join->addBlockToJoin(block);
 
-    if (input.isFinished())
+    if (input.isFinished() && !join->supportParallelJoin())
         join->tryRerangeRightTableData();
 
     set_totals = for_totals;
 }
 
+RerangeRightJoinTableTransform::RerangeRightJoinTableTransform(Block input_header, JoinPtr join_)
+    : IProcessor({input_header}, {Block()})
+    , join(std::move(join_))
+{}
+
+IProcessor::Status RerangeRightJoinTableTransform::prepare()
+{
+    auto & input = inputs.front();
+    if (input.isFinished())
+    {
+        if (!rerange_finished)
+            return Status::Ready;
+        else
+        {
+            auto & output = outputs.front();
+            output.finish();
+            input.close();
+            return Status::Finished;
+        }
+    }
+    else
+    {
+        input.setNeeded();
+        return Status::Ready;
+    }
+}
+
+void RerangeRightJoinTableTransform::work()
```

Comment:
it seems we could avoid creating a dedicated processor just to call this method on finish. could we call it on `input.isFinished()` in `FillingRightJoinSideTransform` instead?

Reply:
if do it in `FillingRightJoinSideTransform`, it would encounter data race problems, as the grace hash join `supportParallelJoin`  is true,  means that `FillingRightJoinSideTransform` run parallelly, but we only need to rerange right table once, it seems that not suitable to do the rerange. So we build another processor transform, which is resized by 1, not run parallelly, and to rerange the right table.

Reply:
we have similar situation in aggregation for example, when we need to trigger merging only once after pre-aggregation phase is finished. it is solved by this questionable pattern 

https://github.com/ClickHouse/ClickHouse/blob/ce82c19c2c4c954f5685453bd48b89f1a17ef6d5/src/Processors/Transforms/AggregatingTransform.cpp#L728-L734

Reply:
In `FillingRightJoinSideTransform` it can not detect whether other input is finished,  so we do not know when to rerange the table in `FillingRightJoinSideTransform`. And I simplify code by move the rerange table logic to `joinBlock` of `GraceHashJoin` and make it rerange only once by use a `hash_join_mutex` lock.